src/annotator/util/buckets.js

```diff
@@ -88,161 +95,151 @@ export function findClosestOffscreenAnchor(anchors, direction) {
+      aboveScreen.add(anchorPosition.anchor);
+      return;
+    } else if (
+      anchorPosition.top >
+      window.innerHeight - BUCKET_BOTTOM_THRESHOLD
+    ) {
+      belowScreen.add(anchorPosition.anchor);
+      return;
+    }
 
+    if (!currentBucket) {
+      // We've encountered our first on-screen anchor position:
+      // We'll need a bucket!
+      currentBucket = _nextBucket(buckets, null, anchorPosition);
+      return;
+    }
+    // We want to contain overlapping highlights and those near each other
+    // within a shared bucket
+    const isContainedWithin =
+      anchorPosition.top > currentBucket.top &&
+      anchorPosition.bottom < currentBucket.bottom;
+
+    // The new anchor's position is far enough below the bottom of the current
+    // bucket to justify starting a new bucket
     const isLargeGap =
-      buckets.length &&
-      position - buckets[buckets.length - 1].position > BUCKET_GAP_SIZE;
-
-    if (current.anchors.length === 0 || isFirstOrLastPoint || isLargeGap) {
-      // Create a new bucket, because:
-      // - There are no more open/working anchors, OR
-      // - This is the first or last point, OR
-      // - There's been a large dimensional gap since the last bucket's position
-      buckets.push({ anchors: current.anchors.slice(), position });
+      anchorPosition.top - currentBucket.bottom > BUCKET_GAP_SIZE;
+
+    if (isLargeGap && !isContainedWithin) {
+      // We need to start a new bucket; push the working bucket and create
+      // a new bucket
+      currentBucket = _nextBucket(buckets, currentBucket, anchorPosition);
     } else {
-      // Merge bucket contents
-
-      // Always remove the last bucket
-      const ultimateBucket = buckets.pop() || /** @type Bucket */ ({});
-
-      // Merge working anchors into the last bucket's anchors
-      let mergedAnchors = [...ultimateBucket.anchors, ...current.anchors];
-      let mergedPosition = ultimateBucket.position;
-
-      // If there is a previous bucket (penultimate bucket) and it has anchors
-      // in it (is not empty)
-      if (buckets[buckets.length - 1]?.anchors.length) {
-        // Remove the previous bucket, too
-        const penultimateBucket = /** @type Bucket */ (buckets.pop() || {});
-        // Merge the penultimate bucket's anchors into our working set of
-        // merged anchors
-        mergedAnchors = [...penultimateBucket.anchors, ...mergedAnchors];
-        // We'll use the penultimate bucket's position as the position for
-        // the merged bucket
-        mergedPosition = penultimateBucket.position;
-      }
-
-      // Push the now-merged bucket onto the buckets collection
-      buckets.push({
-        anchors: Array.from(new Set(mergedAnchors)), // De-dupe anchors
-        position: mergedPosition,
-      });
+      // We'll add this anchor to the current working bucket and update
+      // offset properties accordingly
+      const updatedTop =
```

Comment:
I'm hoping that typedef documentation is clear enough to not warrant line-by-line commenting here, but I can add it!